framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,229 @@
+        NdpOptionType::SourceLinkLayerAddress
+    }
+}
+
+/// Provides functions to implementors that allow the implementor to scan their NDP message for
+/// NDP options.  The type of options available are specific to the type of NDP Message
+/// Ensures that implementors support NdpMessageContents, which further enforces EndOffset + Sized
+/// which we require to do parsing
+pub trait NdpOptions: NdpMessageContents {
+    /// Attempts to lookup the source link layer address from the underlying NDP Message
+    fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let opt = self.option_scan::<LinkLayerAddressOption>(payload_length);
+        opt.map(|sll| &sll.addr)
+    }
+
+    /// Attempts to lookup the mtu option from the underlying NDP Message
+    fn get_mtu_option(&self, payload_length: u16) -> Option<u32> {
+        let mtu_option = self
+            .option_scan::<MtuOption>(payload_length)
+            .map(|mtu| mtu.get_mtu());
+        mtu_option
+    }
+
+    /// Searches through a packet for an option, returns the first it finds or none
+    /// The underlying type is determine by implementing the NdpOption trait which returns the NDP Option Type
+    fn option_scan<T: NdpOption>(&self, payload_length: u16) -> Option<&T> {
+        unsafe {
+            // the start of the options we offset from is the start of the ndp message
+            let opt_start = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset();
+            let requested_type = T::get_type();
+
+            // make sure we do not scan beyond the end of the packet
+            while (payload_offset as u16) < payload_length {
+                // we want to pull the first two bytes off, option type and option length
+                let seek_to = opt_start.add(payload_offset);
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                // first option meta is the option type, let's cast it to our NDPOptionType
+                let cur_type: NdpOptionType = FromPrimitive::from_u8(option_meta[0]).unwrap();
+
+                // second option is the length in octets (8 byte chunks).  So a length of 1 means 8 bytes
+                let cur_size = option_meta[1] * 8;
+
+                if cur_size == 0 {
+                    // Don't know how we will be here, but it is an error condition as we will
+                    // go into an infinite loop and blow the stack if we skip, so exit None
+                    warn!("Option length is set to zero for option type {}", cur_type);
+                    return None;
+                } else if (payload_offset + cur_size as usize) > (payload_length as usize) {
+                    // we need to protect ourselves in case the option_length goes past the payload length
+                    warn!("Option length exceeds the payload length for option type {} option_length {}", cur_type, cur_size);
```

Comment:
so there's also an assumption being made here that `cur_size` is the same as size of `T`, which is not guaranteed on malformed packets. I think we care less about if `payload_offset + cur_size > payload_length`, the next iteration of the while loop will take care of that. It's `payload_offset + sizeof(T) > payload_length` what we should guard against, that's when we will start overlaying into junk memory if not checked.

Reply:
if we had a malformed packet, we should fail it b/c the checksum won't be correct

Reply:
it will still be possible to have a packet that passes the checksum while still have inconsistent type code/length combination in the packet. how likely is this I don't know, but since we are dealing with bytes received, we can't control that.

the main reason I believe we have to deal with this here is because we are doing unsafe code with pointers. if we don't check for size bounds properly and end up touching memory we don't own, all kinds of weird things can happen. even potentially end up with a segfault that crashes the process.

Reply:
i think we are covered in that situation. i've added a test case here https://github.com/williamofockham/NetBricks/pull/58/files#diff-20308be922afb631a0a04084c7be965fR130
the option length is > packet length in that test case. before that check was put in, the process would throw an exception if we exceeded the packet length. we return `None` and stop parsing options.

Reply:
yeah so there are 3 values here.

* how much buffer is left to be parsed, `payload_length`
* option length that's in the packet, `option_length`
* the actual size of T, that's based on the rust type definition, `struct_length`

what's being checked here is `option_length <= payload_length`, but that's not the right check because we are overlaying `T` over the rest of the payload, and we can't assume that `struct_length == option_length`. they should be in all cases, but what if they are not? so the right check is `struct_length <= payload_length`.